casual/casual-inbound-api/src/main/java/se/laz/casual/jca/inflow/CasualInboundTransactionRegistry.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2025, The casual project. All rights reserved.
+ *
+ * This software is licensed under the MIT license, https://opensource.org/licenses/MIT
+ */
+
+package se.laz.casual.jca.inflow;
+
+import io.netty.channel.ChannelId;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Logger;
+
+public class CasualInboundTransactionRegistry
+{
+    private static final Logger log = Logger.getLogger(CasualInboundTransactionRegistry.class.getName());
+    private static final String CHANNEL_ID_CAN_NOT_BE_NULL = "channelId can not be null";
+    private final Map<ChannelId, Set<XidKey>> transactions = new ConcurrentHashMap<>();
+
+    public void add(ChannelId channelId, XidKey key)
+    {
+        Objects.requireNonNull(channelId, CHANNEL_ID_CAN_NOT_BE_NULL);
+        Objects.requireNonNull(key, "key can not be null");
+        log.finest(() -> "adding transaction " + key + " for channel id: " + channelId + " to inbound transaction registry");
+        transactions.computeIfAbsent(channelId, id -> ConcurrentHashMap.newKeySet()).add(key);
+    }
+    public void remove(ChannelId channelId, XidKey key)
+    {
+        Objects.requireNonNull(channelId, CHANNEL_ID_CAN_NOT_BE_NULL);
+        Objects.requireNonNull(key, "key can not be null");
+        log.finest(() -> "removing transaction " + key + "for channel id: " + channelId +" from transaction registry");
+        transactions.computeIfPresent(channelId, (id, set) ->{
```

Comment:
I think we have a race condition here.

If add is called concurrently to remove, after the set.isEmpty check, something is added, but we have then placed null rather than the set back.
Could we prevent this by just leaving the empty set in place rather than removing, then if something is later added ot the channel it will just use the empty set to add the next one.
Yes a bit of memory overhead, but it will remove this race and also slightly reduce the add performance as we won't need to make a new set each time the channels transactions are all completed, which i think will likely be a common situations.
run a few transactions, idle, run a few more, etc.